hoomd/BoxResizeUpdater.cc

```diff
@@ -110,52 +111,46 @@ void BoxResizeUpdater::update(unsigned int timestep)
-                // positions without failing the check)
-                Scalar3 scaled_pos = new_box.makeCoordinates(fractional_pos);
-                h_pos.data[i].x = scaled_pos.x;
-                h_pos.data[i].y = scaled_pos.y;
-                h_pos.data[i].z = scaled_pos.z;
-                }
-            }
+        // move the particles to be inside the new box
+        ArrayHandle<Scalar4> h_pos(m_pdata->getPositions(),
+                                   access_location::host,
+                                   access_mode::readwrite);
+
+        unsigned int n_particle_group = m_scale_particles->getNumMembersGlobal();
+
+        for (unsigned int i = 0; i < n_particle_group; i++)
+        {
+        // obtain scaled coordinates in the old global box
+        Scalar3 fractional_pos = cur_box.makeFraction(
+        make_scalar3(h_pos.data[i].x,
+                     h_pos.data[i].y,
+                     h_pos.data[i].z));
+
+        // intentionally scale both rigid body and free particles, this
+        // may waste a few cycles but it enables the debug inBox checks
+        // to be left as is (otherwise, setRV cannot fixup rigid body
+        // positions without failing the check)
+        Scalar3 scaled_pos = new_box.makeCoordinates(fractional_pos);
+        h_pos.data[i].x = scaled_pos.x;
+        h_pos.data[i].y = scaled_pos.y;
+        h_pos.data[i].z = scaled_pos.z;
+        }
+
         // otherwise, we need to ensure that the particles are still in their
         // local boxes by wrapping them if they are not
-        else
-            {
-            ArrayHandle<Scalar4> h_pos(m_pdata->getPositions(),
-                                       access_location::host,
-                                       access_mode::readwrite);
-
-            ArrayHandle<int3> h_image(m_pdata->getImages(),
-                                      access_location::host,
-                                      access_mode::readwrite);
-
-            const BoxDim& local_box = m_pdata->getBox();
-
-            for (unsigned int i = 0; i < m_pdata->getN(); i++)
-                {
-                // need to update the image if we move particles from one side
-                // of the box to the other
-                local_box.wrap(h_pos.data[i], h_image.data[i]);
-                }
-            }
+        ArrayHandle<int3> h_image(m_pdata->getImages(),
+                                  access_location::host,
+                                  access_mode::readwrite);
+
+        const BoxDim& local_box = m_pdata->getBox();
+
+        for (unsigned int i = 0; i < m_pdata->getN(); i++)
+          {
+          // need to update the image if we move particles from one side
+          // of the box to the other
+          local_box.wrap(h_pos.data[i], h_image.data[i]);
+          }
+          //TODO: Is it faster/better to do it like this, or one loop and if statement with isMember
```

Comment:
I also tried something like:

```C++
for (unsigned int i=0;i<m_pdata->getN(); i++)
{
if (m_scale_particles->isMember(i))
    {
    // scale particles
    }
    else
    {
    // update image
    }
}
```

I tried to run this earlier and it seemed slower than the current set up, but I can time both to see if one performs better -- or if there is a better way to set up this code.

Reply:
Either should be fine. `isMember` is an O(1) check. The two loops method only adds an extra cost of a for loop over a few more indices. However, it would be great to test both and confirm.

Note that in this implementation your first loop is missing a group member lookup. This code from `TwoStepLangevin.cc` line 83 shows how it should look:
```
    for (unsigned int group_idx = 0; group_idx < group_size; group_idx++)
        {
        unsigned int j = m_group->getMemberIndex(group_idx);
```

Reply:
The timed trials seem to confirm that there is no meaningful difference between the two ways to write this code:
![image](https://user-images.githubusercontent.com/47036428/107572856-c8176e80-6bba-11eb-84ac-5ea39ab1919c.png)
![image](https://user-images.githubusercontent.com/47036428/107572869-cc438c00-6bba-11eb-906f-fbd19a1d8056.png)

Timed trials using "one loop" (used `m_scale_particles->isMember`) and two loops (one for to scale group members and one to wrap all points). 

I'm not sure why the run time increases sharply at 30k points for both methods.

Reply:
Most likely 30k points exceeds the size of the fastest cache. You can see two other slopes in the plots as the performance scales with the next fastest cache up to 125k and then with DRAM speeds beyond that.

Reply:
That's pretty cool. Thanks for the explanation!